repo-updater: add endpoint to schedule permissions sync

unknwon opened the pull request:
This PRs adds a new endpoint to repo-updater for scheduling both user and repository permissions sync. 

It is step one of three steps for https://github.com/sourcegraph/sourcegraph/issues/8990.

_Please review by commit._

mrnugget commented on cmd/repo-updater/repoupdater/server_test.go:
```diff
@@ -1149,6 +1149,79 @@ func (g *fakeGitserverClient) ListCloned(ctx context.Context) ([]string, error)
+		name          string
+		permsSyncer   *fakePermsSyncer
+		body          string
+		expStatusCode int
```
Tiniest nitpick of all time, feel free to ignore: maybe use `want` instead of `exp` here? It's one more character, but far clearer at first glance. And we use it in other tests, too.

unknwon replied:
:D Interestingly I also learned using `exp` from other tests, but `want` SGTM 👍

mrnugget commented on internal/repoupdater/protocol/repoupdater.go:
```diff
@@ -170,6 +170,25 @@ type ChangesetSyncResponse struct {
+// PermsSyncRequest is a request to sync permissions.
+type PermsSyncRequest struct {
+	Type PermsSyncType `json:"type"`
+	IDs  []int32       `json:"ids"`
```
Suggestion, no blocker: use two distinct slices, `UserIDs` and `RepoIDs`, which are used depending on the `Type`. That makes it clearer what the IDs are (compared to the`IDs` field) and saves you the casting to `api.RepoID` further down the line.

unknwon replied:
Actually.... I just got inspired by your idea! If we have `UserIDs` and `RepoIDs`, that is clear which type of ids they are, and no need to have `Type` anymore. Also there is no reason why we can't schedule both users and repos at the same time on this API level!

keegancsmith commented on cmd/repo-updater/shared/main.go:
```diff
@@ -202,7 +202,7 @@ func Main(enterpriseInit EnterpriseInit) {
 
 	addr := net.JoinHostPort(host, port)
-	log15.Info("server listening", "addr", addr)
+	log15.Info("repo-updater: listening", "addr", addr)
```
minor: why is this an improvement? In dev and single server image the service name is already prefixed. In kubernetes the logs will be attached to the pod.

unknwon replied:
I was looking at the log and trying to find which port repo-updater was listening on, and couldn't spot for a minute, then I realize it is the only service that does not spell out its name along with the listening address:

```
17:57:21            frontend | ✱ Sourcegraph is ready at: http://localhost:3080
17:57:21        query-runner | INFO query-runner: listening, addr: 127.0.0.1:3183
17:57:22        github-proxy | INFO github-proxy: listening, addr: 127.0.0.1:3180
17:57:22            replacer | INFO replacer: listening, addr: 127.0.0.1:3185
17:57:23            searcher | INFO searcher: listening, addr: 127.0.0.1:3181
17:57:25           gitserver | INFO git-server: listening, addr: 127.0.0.1:3178
17:57:25        repo-updater | INFO repo-updater: listening, addr: 127.0.0.1:3182
17:57:25             symbols | INFO symbols: listening, addr: 127.0.0.1:3184
```

keegancsmith commented on cmd/repo-updater/repoupdater/server_test.go:
```diff
@@ -1149,6 +1149,79 @@ func (g *fakeGitserverClient) ListCloned(ctx context.Context) ([]string, error)
+
+			if w.Code != test.expStatusCode {
+				t.Fatalf("Code: want %v but got %v", test.expStatusCode, w.Code)
+			} else if diff := cmp.Diff(w.Body.String(), test.expBody); diff != "" {
+				t.Fatalf("Body: %v", diff)
```
minor: useful to have want first and got as the second param. Also useful to indicate in the error message which is want and got and put the diff on its own line.

```suggestion
			} else if diff := cmp.Diff(test.expBody, w.Body.String()); diff != "" {
				t.Fatalf("Body mismatch (-want +got):\n%s", diff)
```

unknwon commented on cmd/repo-updater/repos/scheduler.go:
```diff
@@ -401,17 +401,20 @@ type updateQueue struct {
-type priority int
+// Priority defines how urgent the syncing or update request is. Generally,
+// if the request is driven from a user action, then it should be PriorityHigh.
+// All other cases should be PriorityLow.
```
Technically yes :D But may not be necessary?

unknwon commented:
@keegancsmith 

> Minor discussion point: You made priority public for this PR, but you end up only ever passing in HighPriority via the endpoint. Alternatively I think you could of just exported a function which only does high priority enqueuing? This could of been named appropriately and would of avoided the need to understand priorities from the exported API.

The RPC handler added in this PR turns out to be the only call site, so I reverted changes of making `repos.prioiry` public and updated docstring of two methods involved. PTAL!

> Meta: for reviewing commit by commit it is still useful to explain the why in the commit message body. eg why did you make Priority field public? Its clearer once you have reviewed the whole PR, but when reviewing commit by commit it makes it clear as you review the commit.

👍 Thanks for the suggestion! Isn't a heavy user of commit message body, will do that to make review easier next time doing by-commit PRs :)
